composer/loggers/slack_logger.py

```diff
@@ -0,0 +1,219 @@
+        Otherwise, wait for the next log_interval (batch end or epoch end) to flush the buffer.
+        """
+        filtered_data = {k: v for k, v in data.items() if re.match(self.regex_all_keys, k) is not None}
+        self.log_dict.update(filtered_data)
+
+        if len(self.log_dict.keys()) >= self.max_logs_per_message:
+            self._flush_logs_to_slack(**kwargs)
+
+    def _default_log_bold_key_normal_value_pair_with_header(self, data: Dict[str, Any],
+                                                            **kwargs) -> List[Dict[str, Any]]:
+        """Default formatter function if no formatter func is specified.
+
+        This function will:
+        1. Log the key-value pairs in bold (key) and normal (value) text.
+        2. When logging metrics, set the step number as the header of the section.
+
+        Args:
+            data (Dict[str, Any]): Data to be logged.
+            **kwargs: Additional arguments to be passed to the formatter function (eg header)
+
+        Returns:
+            List[Dict[str, str]]: List of blocks to be sent to Slack.
+        """
+        blocks = [{'type': 'section', 'text': {'type': 'mrkdwn', 'text': f'*{k}:* {v}'}} for k, v in data.items()]
+        if len(blocks) > 0 and 'header' in kwargs:
+            header = kwargs['header']
+            blocks.append({'type': 'header', 'text': {'type': 'plain_text', 'text': f'{header}'}})
+
+        return blocks
+
+    def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None) -> None:
+        self._log_to_buffer(data=metrics, header=step)
+
+    def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
+        self._log_to_buffer(data=hyperparameters)
+
+    def log_traces(self, traces: Dict[str, Any]):
+        self._log_to_buffer(data=traces)
+
+    def epoch_end(self, state: State, logger: Logger) -> None:
+        cur_epoch = int(state.timestamp.epoch)  # epoch gets incremented right before EPOCH_END
+        unit = self.log_interval.unit
+
+        if unit == TimeUnit.EPOCH and (cur_epoch % int(self.log_interval) == 0 or cur_epoch == 1):
+            self._flush_logs_to_slack()
+
+    def batch_end(self, state: State, logger: Logger) -> None:
+        cur_batch = int(state.timestamp.batch)
+        unit = self.log_interval.unit
+        if unit == TimeUnit.BATCH and (cur_batch % int(self.log_interval) == 0 or cur_batch == 1):
+            self._flush_logs_to_slack()
+
+    def close(self, state: State, logger: Logger) -> None:
+        self._flush_logs_to_slack()
+
+    def _flush_logs_to_slack(self, **kwargs) -> None:
+        """Flush buffered metadata to MosaicML.
+
+        Format slack messages through rich message layouts created using Slack Blocks Kit.
+        See documentation here: https://api.slack.com/messaging/composing/layouts.
+        """
+        channel_id = self.channel_id
+        if channel_id is None:
+            raise TypeError
+
+        inx = 0
+        while inx < len(self.log_dict.keys()):
```

Comment:
this just sends max_logs_per_message blocks at a time, but it still sends all the messages?

Reply:
if there's 1000 metrics in between batches, this will flush all 1000 metrics 50 at a time